x/emissions/keeper/actor_penalties.go

```diff
@@ -0,0 +1,126 @@
+			if err != nil {
+				return false, errorsmod.Wrap(err, "error getting inferer score EMA")
+			}
+			scores = append(scores, score.Score)
+			return false, nil
+		},
+	); err != nil {
+		return errorsmod.Wrap(err, "error walking active inferers")
+	}
+
+	params, err := k.GetParams(ctx)
+	if err != nil {
+		return errorsmod.Wrap(err, "error getting params")
+	}
+	standardDeviation, _, err := alloraMath.MedianAbsoluteDeviation(scores)
+	if err != nil {
+		return errorsmod.Wrap(err, "error calculating median absolute deviation")
+	}
+	x, err := standardDeviation.Mul(params.SortitionLambdaPenalty)
+	if err != nil {
+		return errorsmod.Wrap(err, "error calculating penalty")
+	}
+	penalty, err := lowestEMAScore.Score.Sub(x)
+	if err != nil {
+		return errorsmod.Wrap(err, "error calculating penalty")
+	}
+
+	return k.penaltyInTopic.Set(ctx, topicId, penalty)
+}
+
+// MayPenaliseInferer penalises an inferer for missing previous epochs. It saves and returns the new EMA score.
+// If the inferer didn't miss any epochs this is a no-op, the EMA score is returned as is.
+func (k *Keeper) MayPenaliseInferer(
+	ctx sdk.Context,
+	topic types.Topic,
+	block types.BlockHeight,
+	emaScore types.Score,
+) (types.Score, error) {
+	missedEpochs := countWorkerContiguousMissedEpochs(topic, emaScore.BlockHeight)
+	// No missed epochs == no penalty
+	if missedEpochs == 0 {
+		return emaScore, nil
+	}
+
+	penalty, err := k.penaltyInTopic.Get(ctx, topic.Id)
+	// No penalty set, no penalty
+	if errors.Is(err, collections.ErrNotFound) {
+		return emaScore, nil
+	}
+	if err != nil {
+		return types.Score{}, err
+	}
+	emaScore.BlockHeight = block
+	// TODO: Provide the limit as the initial score
+	emaScore.Score, err = applyPenalty(topic, penalty, alloraMath.ZeroDec(), emaScore.Score, missedEpochs)
+	if err != nil {
+		return types.Score{}, err
+	}
+
+	// Save the penalised EMA score
+	return emaScore, k.SetInfererScoreEma(ctx, topic.Id, emaScore.Address, emaScore)
+}
+
+// applyPenalty applies the penalty to the EMA score for the given number of missed epochs while staying above provided limit.
+func applyPenalty(topic types.Topic, penalty, limit, emaScore alloraMath.Dec, missedEpochs uint32) (alloraMath.Dec, error) {
+	for i := uint32(0); i < missedEpochs; i++ {
+		penalisedScore, err := alloraMath.CalcEma(topic.MeritSortitionAlpha, penalty, emaScore, false)
+		if err != nil {
+			return alloraMath.ZeroDec(), err
+		}
+
+		if penalisedScore.Lt(limit) {
+			break
+		}
+		emaScore = penalisedScore
+	}
+
+	return emaScore, nil
+}
+
+// CountWorkerContiguousMissedEpochs counts the number of contiguous missed epochs prior to the given nonce, given the
+// last worker submission and the current block heights.
+func countWorkerContiguousMissedEpochs(topic types.Topic, lastSubmissionHeight int64) uint32 {
+	count := uint32(0)
+	prevEpochStart := topic.EpochLastEnded - topic.EpochLength
+	for lastSubmissionHeight < prevEpochStart {
```

Comment:
Could this be more efficient with a division? 

Like 

```
    if lastSubmissionHeight >= prevEpochStart {
        return 0
    }

    // Calculate missed epochs using division
    // Add 1 to match the original implementation's counting behavior
    missedEpochs := ((prevEpochStart - lastSubmissionHeight) / topic.EpochLength) + 1
    
    return uint32(missedEpochs)
```

Reply:
Definitely, this was bad :)